### PR TITLE
Fix MDC issues

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Ensure that the Log4j MDC context is always propagated. See the commit messages for details.

### CODE REVIEW

File: plugins/commands/advisor/src/main/kotlin/AdvisorCommand.kt [hunk 1]
```diff
@@ -125,6 +125,8 @@ class AdvisorCommand : OrtCommand(
         val advisor = Advisor(distinctProviders, ortConfig.advisor)
 
         val ortResultInput = readOrtResult(ortFile)
+
+        @Suppress("ForbiddenMethodCall")
```
Comment (sschuberth) [inline plugins/commands/advisor/src/main/kotlin/AdvisorCommand.kt:129]: Please also explain in the commit message in which cases the new rule is suppressed / it is ok to suppress the rule.

File: .detekt.yml [hunk 2]
```diff
@@ -70,6 +70,11 @@ style:
     maxDestructuringEntries: 4
   ForbiddenComment:
     active: false
+  ForbiddenMethodCall:
+    active: true
+    # Use org.ossreviewtoolkit.utils.ort.runBlocking instead in all code that can be used as a library.
+    # One exception to that are the CLI commands.
```
Comment (sschuberth) [inline .detekt.yml:76]: IMO that comment better goes to the commit message, see my comment below.

File: .detekt.yml [hunk 3]
```diff
@@ -70,6 +70,11 @@ style:
     maxDestructuringEntries: 4
   ForbiddenComment:
     active: false
+  ForbiddenMethodCall:
+    active: true
+    # Use org.ossreviewtoolkit.utils.ort.runBlocking instead in all code that can be used as a library.
```
Comment (sschuberth) [inline .detekt.yml:75]: IMO this line can be dropped in favor of keeping the per-entry comment in line 77.

File: scanner/src/main/kotlin/provenance/NestedProvenanceResolver.kt [hunk 4]
```diff
@@ -38,7 +36,7 @@ interface NestedProvenanceResolver {
      * [NestedProvenance] always contains only the provided [ArtifactProvenance]. For a [RepositoryProvenance] the
      * resolver looks for nested repositories, for example Git submodules or Mercurial subrepositories.
      */
-    fun resolveNestedProvenance(provenance: KnownProvenance): NestedProvenance
+    suspend fun resolveNestedProvenance(provenance: KnownProvenance): NestedProvenance
```
Comment (sschuberth) [inline scanner/src/main/kotlin/provenance/NestedProvenanceResolver.kt:39]: Can callers like

https://github.com/oss-review-toolkit/ort/blob/74557ba1428f578d1456535435591292714c469d/scanner/src/main/kotlin/Scanner.kt#L272-L276

now be simplified?

File: scanner/src/main/kotlin/provenance/PackageProvenanceResolver.kt [hunk 5]
```diff
@@ -52,7 +50,7 @@ interface PackageProvenanceResolver {
      *
      * Throws an [IOException] if the provenance cannot be resolved.
      */
-    fun resolveProvenance(pkg: Package, defaultSourceCodeOrigins: List<SourceCodeOrigin>): KnownProvenance
+    suspend fun resolveProvenance(pkg: Package, defaultSourceCodeOrigins: List<SourceCodeOrigin>): KnownProvenance
```
Comment (sschuberth) [inline scanner/src/main/kotlin/provenance/PackageProvenanceResolver.kt:53]: Can callers like

https://github.com/oss-review-toolkit/ort/blob/74557ba1428f578d1456535435591292714c469d/scanner/src/main/kotlin/Scanner.kt#L245-L249

now be simplified?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/provenance/NestedProvenanceResolver.kt (hunk 4 above)
Comment (mnonnenmacher) [reply scanner/src/main/kotlin/provenance/NestedProvenanceResolver.kt:39]: No, the `async` is used here to run multiple resolutions in parallel, the `runCatching` to catch exceptions, and the `provenance to` to make the `provenance` available for later code.

File: scanner/src/main/kotlin/provenance/PackageProvenanceResolver.kt (hunk 5 above)
Comment (mnonnenmacher) [reply scanner/src/main/kotlin/provenance/PackageProvenanceResolver.kt:53]: No, the `async` is used here to run multiple resolutions in parallel, the `runCatching` to catch exceptions, and the `pkg to` to make the `pkg` available for later code.

File: .detekt.yml (hunk 3 above)
Comment (mnonnenmacher) [reply .detekt.yml:75]: I wanted to provide a bit more detail here, I thought someone might first look for an explanation here before searching the Git history. But I forgot to remove the comment in line 77.

File: .detekt.yml (hunk 2 above)
Comment (mnonnenmacher) [reply .detekt.yml:76]: See above.

**Comment by mnonnenmacher** [comment]
The errors in the `ClearlyDefinedServiceFunTest` seem to be unrelated to my changes, I get `SocketTimeoutException`s also when I execute them on the main branch. It's a bit unfortunate that they occur now.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**